File: modules/snoozepulse-audio/android/src/main/java/expo/modules/snoozepulseaudio/PcmRingBuffer.kt
package expo.modules.snoozepulseaudio

/**
 * Fixed-capacity PCM ring — allocated once, never resized.
 * Capture thread writes; DSP / snippet path reads the most recent samples without allocating.
 */
internal class PcmRingBuffer(private val capacity: Int) {
  private val data = ShortArray(capacity)
  private var writeIndex = 0
  private var filled = 0

  val length: Int
    get() = filled

  /** Copy [src] into the ring. Caller must not pass a newly allocated buffer each frame. */
  @Synchronized
  fun write(src: ShortArray, offset: Int, count: Int) {
    var remaining = count
    var srcPos = offset
    while (remaining > 0) {
      val chunk = minOf(remaining, capacity - writeIndex)
      System.arraycopy(src, srcPos, data, writeIndex, chunk)
      writeIndex = (writeIndex + chunk) % capacity
      srcPos += chunk
      remaining -= chunk
      filled = minOf(capacity, filled + chunk)
    }
  }

  /**
   * Copy the newest [count] samples into [dest] (must be pre-sized).
   * Returns how many samples were copied.
   */
  @Synchronized
  fun copyLatest(count: Int, dest: ShortArray): Int {
    val n = minOf(count, filled, dest.size)
    if (n == 0) {
      return 0
    }
    val start = (writeIndex - n + capacity) % capacity
    val first = minOf(n, capacity - start)
    System.arraycopy(data, start, dest, 0, first)
    if (first < n) {
      System.arraycopy(data, 0, dest, first, n - first)
    }
    return n
  }

  @Synchronized
  fun clear() {
    writeIndex = 0
    filled = 0
  }
}

File: modules/snoozepulse-audio/ios/PcmRingBuffer.swift
import Foundation

/// Fixed-capacity PCM ring — allocated once, never resized on the audio thread.
final class PcmRingBuffer {
  private let capacity: Int
  private var data: [Int16]
  private var writeIndex = 0
  private var filled = 0
  private let lock = NSLock()

  init(capacity: Int) {
    self.capacity = capacity
    self.data = [Int16](repeating: 0, count: capacity)
  }

  var length: Int {
    lock.lock()
    defer { lock.unlock() }
    return filled
  }

  func write(_ src: UnsafePointer<Int16>, count: Int) {
    lock.lock()
    defer { lock.unlock() }
    var remaining = count
    var srcPos = 0
    while remaining > 0 {
      let chunk = min(remaining, capacity - writeIndex)
      for i in 0..<chunk {
        data[writeIndex + i] = src[srcPos + i]
      }
      writeIndex = (writeIndex + chunk) % capacity
      srcPos += chunk
      remaining -= chunk
      filled = min(capacity, filled + chunk)
    }
  }

  /// Copy the newest `count` samples into `dest`. Returns samples copied.
  @discardableResult
  func copyLatest(count: Int, into dest: inout [Int16]) -> Int {
    lock.lock()
    defer { lock.unlock() }
    let n = min(count, min(filled, dest.count))
    if n == 0 { return 0 }
    let start = (writeIndex - n + capacity) % capacity
    for i in 0..<n {
      dest[i] = data[(start + i) % capacity]
    }
    return n
  }

  func clear() {
    lock.lock()
    defer { lock.unlock() }
    writeIndex = 0
    filled = 0
  }
}
